Support adding new keychains

lox opened the pull request:
This pull request adds some new APIs to address #1:

- `NewKeychain(path, password string, promptUser bool)`
- `Keychain.Delete()`
- `Item.SetMatchSearchList(karr ...*Keychain)`
- `Item.UseKeychain(kc *Keychain)`

These match the original keychain API as closely as possible. I considered adding methods on `Keychain` for `AddGenericItem` and `GetGenericItem`, but it looks like those are effectively deprecated? 

Feel free to provide stylistic feedback, objective-c is still new ground for me.

lox commented on keychain.go:
```diff
@@ -125,10 +130,87 @@ var matchTypeRef = map[MatchLimit]C.CFTypeRef{
+		errCode = C.SecKeychainCreate(pathRef, C.UInt32(len(password)), unsafe.Pointer(passwordRef), C.Boolean(0), nil, &kref)
+	}
+
+	// TODO: Without passing in kref I get 'One or more parameters passed to the function were not valid (-50)'
```
This perplexed me, any suggestions @gabriel?

gabriel commented on keychain.go:
```diff
@@ -125,10 +130,87 @@ var matchTypeRef = map[MatchLimit]C.CFTypeRef{
 var ReturnRefKey = attrKey(C.CFTypeRef(C.kSecReturnRef))
 
+// Keychain represents a specific OSX Keychain
+type Keychain struct {
```
Move this to osx.go if it is osx specific?

gabriel commented on keychain.go:
```diff
@@ -125,10 +130,87 @@ var matchTypeRef = map[MatchLimit]C.CFTypeRef{
+
+// NewKeychain creates a new keychain file with either a password, or a triggered prompt to the user
+func NewKeychain(path, password string, promptUser bool) (*Keychain, error) {
+	pathRef := C.CString(path)
```
Use StringToCFString

gabriel replied:
Ditto everywhere else

lox replied:
Is that possible given that `SecKeychainCreate` requires a `*C.char`?

gabriel commented on keychain.go:
```diff
@@ -125,10 +130,87 @@ var matchTypeRef = map[MatchLimit]C.CFTypeRef{
+	path string
+}
+
+var DefaultKeychain *Keychain
```
Don't need this?

gabriel commented on keychain.go:
```diff
@@ -125,10 +130,87 @@ var matchTypeRef = map[MatchLimit]C.CFTypeRef{
+	return C.CFTypeRef(keyRef), err
+}
+
+type keychainArray []*Keychain
```
why not []Keychain?

gabriel commented on keychain.go:
```diff
@@ -125,10 +130,87 @@ var matchTypeRef = map[MatchLimit]C.CFTypeRef{
+
+type keychainArray []*Keychain
+
+func (ka keychainArray) Convert() (C.CFTypeRef, error) {
```
If the person needs to release the returned value should have a comment saying so... ditto for other places

gabriel commented on keychain.go:
```diff
@@ -125,10 +130,87 @@ var matchTypeRef = map[MatchLimit]C.CFTypeRef{
+	var refs []C.CFTypeRef = make([]C.CFTypeRef, len([]*Keychain(ka)))
+	var err error
+
+	for idx, kc := range ka {
```
can you use append and then you dont need an array of pointers

gabriel commented on keychain.go:
```diff
@@ -125,10 +130,87 @@ var matchTypeRef = map[MatchLimit]C.CFTypeRef{
 	MatchLimitAll: C.CFTypeRef(C.kSecMatchLimitAll),
 }
 
+var MatchSearchList = attrKey(C.CFTypeRef(C.kSecMatchSearchList))
```
is this a  key? if so MatchSearchListKey

gabriel commented:
Sorry for lateness in reviewing

lox commented:
Thanks @gabriel, will address these asap.

lox commented:
Apologies, asap ended up being a lot longer than intended. I've updated this and rebased master. I wasn't able to use `StringToCFString` as suggested, as many of the Keychain API calls require a `*C.char`. Open to suggestions on that one @gabriel, my Obj-c skills are lacking.

lox commented:
Any additional feedback @gabriel?

lox commented:
:(

cjb commented:
@lox oh, sorry about the long delay! I'll make a ticket for getting this reviewed so that we can put aside time to get it done. Thanks for the PR and changes!

lox commented:
No stress, are you guys actively using this library?

gabriel commented:
I merged and did some cleanup mostly de-linting.

I changed the constructors for Keychain to `NewKeychain` and `NewKeychainWithPrompt`, since they are mutually exclusive (password doesn't make sense if prompting).

lox commented:
Makes sense, surprised I didn't do that in the first place. Thanks!
